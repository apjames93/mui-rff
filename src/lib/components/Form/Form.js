/* eslint-disable max-len */
import React, { cloneElement } from 'react';
import PropTypes from 'prop-types';
import { Form as ReactFinalForm } from 'react-final-form';

export const RffWrapper = ({
  children,
  debug,
  initialValues,
  initialValuesEqual,
  keepDirtyOnReinitialize,
  onSubmit,
  subscription,
  validate,
  validateOnBlur,
  mutators,
}) => (
  <ReactFinalForm
    debug={debug}
    initialValues={initialValues}
    initialValuesEqual={initialValuesEqual}
    keepDirtyOnReinitialize={keepDirtyOnReinitialize}
    onSubmit={onSubmit}
    subscription={subscription}
    validate={validate}
    validateOnBlur={validateOnBlur}
    mutators={mutators}
  >
    {({ handleSubmit, submitting, form }) => (
      <form onSubmit={handleSubmit}>
        <>
          {Array.isArray(children) && children.map((child, key) => cloneElement(
            child,
            {
              key,
              submitting,
              form,
            },
          ))}

          {!Array.isArray(children) && cloneElement(
            children,
            {
              submitting,
              form,
            },
          )}
        </>
      </form>
    )}
  </ReactFinalForm>
);
export default RffWrapper;

// https://final-form.org/docs/react-final-form/types/FormProps
RffWrapper.propTypes = {
  /**
   * A render function that is given FormRenderProps, as well as any non-API props passed into the <Form/> component.
   */
  children: PropTypes.node.isRequired,
  /**
   * Function to call when the form is submitted. There are three possible ways to write an onSubmit function:
   */
  onSubmit: PropTypes.func.isRequired,

  /**
   * A callback for debugging that receives the form state and the states of all the fields. It's called on every state change. A typical thing to pass in might be console.log.
   */
  debug: PropTypes.func,
  /**
   * The initial values of your form. These will also be used to compare against the current values to calculate pristine and dirty.
   */
  initialValues: PropTypes.shape(),
  /**
   * A predicate to determine whether or not the initialValues prop has changed, i.e. to know if the form needs to be reinitialized with the new values. Useful for passing in a "deep equals" function if you need to. Defaults to "shallow equals".
   */
  initialValuesEqual: PropTypes.bool,
  /**
   * If true, only pristine values will be overwritten when initialize(newValues) is called. This can be useful for allowing a user to continue to edit a record while the record is being saved asynchronously, and the form is reinitialized to the saved values when the save is successful. Defaults to false.
   */
  keepDirtyOnReinitialize: PropTypes.bool,
  /**
   * An object of the parts of FormState to subscribe to. If a subscription is provided, the <Form/> will only rerender when those parts of form state change.
    If no subscription is provided, it will default to subscribing to all form state changes. i.e. <Form/> will rerender whenever any part of the form state changes.
   * ex : { [string]: boolean }
    */
  subscription: PropTypes.shape(),
  /**
   * ex: (values: FormValues) => Object | Promise<Object>
   * A whole-record validation function that takes all the values of the form and returns any validation errors. There are two possible ways to write a validate function:
   */
  validate: PropTypes.func,
  /**
   * If true, validation will happen on blur. If false, validation will happen on change. Defaults to false.
   */
  validateOnBlur: PropTypes.bool,
  /**
   * Optional.
   * Named mutator functions.
   */
  mutators: PropTypes.shape(),
};

RffWrapper.defaultProps = {
  debug: undefined,
  initialValues: {},
  initialValuesEqual: false,
  keepDirtyOnReinitialize: false,
  subscription: undefined,
  validate: undefined,
  validateOnBlur: false,
  mutators: undefined,
};
