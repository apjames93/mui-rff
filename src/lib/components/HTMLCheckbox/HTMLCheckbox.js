import React from 'react';
import PropTypes from 'prop-types';
import { FormControlLabel } from '@material-ui/core';
import RFFField from '../RFFField/RFFField';

export const HTMLCheckbox = ({
  name,
  label,
  labelPlacement,
}) => (
  <RFFField
    type="checkbox"
    name={name}
    passProps={field => ({
      id: field.input.name,
      value: field.input.value,
      onChange: field.input.onChange,
    })}
  >
    <FormControlLabel
      control={(
        <input type="checkbox" />
      )}
      label={label}
      labelPlacement={labelPlacement}
    />
  </RFFField>

);

export default HTMLCheckbox;

HTMLCheckbox.propTypes = {
  /**
   * The name of your field. Field values may be deeply nested using dot-and-bracket syntax.
   */
  name: PropTypes.string.isRequired,
  /**
   * The label content.
   */
  label: PropTypes.string,
  /**
   * MUI Props: 'bottom' | 'end' | 'start' | 'top'
   * The position of the label.
   */
  labelPlacement: PropTypes.string,
};

HTMLCheckbox.defaultProps = {
  label: '',
  labelPlacement: 'end',
};
