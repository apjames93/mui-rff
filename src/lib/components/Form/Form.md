Form example:

```js noeditor
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';
```

```js
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';


const onSubmit = async (values) => {
  console.log('Input Form', values)
}

  <Form
    onSubmit={onSubmit}
  >
    <MuiInput
      name="thing"
      label="thing"
    />
    <MuiInput
      name="number"
      label="number"
      type="number"
    />
    <MuiInput
      name="date"
      label="date"
      type="date"
    />
    <MuiSelect
      name="stuff"
      options={[{ num: 'six' }, { num: 'nine' }]}
      displayKey="num"
    />
    <MuiSubmit />
  </Form>
```
