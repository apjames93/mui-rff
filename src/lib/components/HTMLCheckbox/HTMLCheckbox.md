HTMLCheckbox example:

```js noeditor
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import HTMLCheckbox from './HTMLCheckbox';
```

```js
import Form from '../Form/Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import HTMLCheckbox from './HTMLCheckbox';


const onSubmit = async (values) => {
  console.log('HTML Checkbox Form', values)
}

  <Form
    onSubmit={onSubmit}
  >
    <HTMLCheckbox
      name="favoriteSport"
      label="Sports"
      labelPlacement="top"
    />
    <MuiSubmit />
  </Form>
```
