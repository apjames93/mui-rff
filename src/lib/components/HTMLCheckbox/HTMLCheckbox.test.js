import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { HTMLCheckbox } from './HTMLCheckbox';
import Form from '../Form/Form';

function setup() {
  const props = {
    name: 'cool',
    label: 'apples',
  };
  const comp = shallow(<HTMLCheckbox {...props} />);
  return { comp, props };
}

describe('<HTMLCheckbox />', () => {
  it('renders Input', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(
      <Form onSubmit={() => true}>
        <MemoryRouter>
          <HTMLCheckbox {...props} />
        </MemoryRouter>
      </Form>,
    );
    expect(tree).toMatchSnapshot();
  });
});
