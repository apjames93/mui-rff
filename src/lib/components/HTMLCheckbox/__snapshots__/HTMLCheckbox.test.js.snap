// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HTMLCheckbox /> snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    className="MuiFormControlLabel-root"
    field={
      Object {
        "input": Object {
          "checked": false,
          "name": "cool",
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "type": "checkbox",
          "value": null,
        },
        "meta": Object {
          "active": false,
          "data": Object {},
          "dirty": false,
          "dirtySinceLastSubmit": false,
          "error": undefined,
          "initial": false,
          "invalid": false,
          "length": undefined,
          "modified": false,
          "modifiedSinceLastSubmit": false,
          "pristine": true,
          "submitError": undefined,
          "submitFailed": false,
          "submitSucceeded": false,
          "submitting": false,
          "touched": false,
          "valid": true,
          "validating": false,
          "visited": false,
        },
      }
    }
    id="cool"
  >
    <input
      onChange={[Function]}
      type="checkbox"
      value={null}
    />
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      apples
    </span>
  </label>
</form>
`;
